sync15-adapter/src/request.rs

```diff
@@ -0,0 +1,1203 @@
+    pub max_request_bytes: usize,
+
+    /// The maximum number of records that can be uploaded to a collection in a single POST request.
+    #[serde(default = "usize::max_value")]
+    pub max_post_records: usize,
+
+    /// The maximum combined size in bytes of the record payloads that can be uploaded to a
+    /// collection in a single POST request.
+    #[serde(default = "usize::max_value")]
+    pub max_post_bytes: usize,
+
+    /// The maximum total number of records that can be uploaded to a collection as part of a
+    /// batched upload.
+    #[serde(default = "usize::max_value")]
+    pub max_total_records: usize,
+
+    /// The maximum total combined size in bytes of the record payloads that can be uploaded to a
+    /// collection as part of a batched upload.
+    #[serde(default = "usize::max_value")]
+    pub max_total_bytes: usize,
+
+    /// The maximum size of an individual BSO payload, in bytes.
+    #[serde(default = "default_max_record_payload_bytes")]
+    pub max_record_payload_bytes: usize,
+}
+
+// This is annoying but seems to be the only way to do it...
+fn default_max_request_bytes() -> usize { 260 * 1024 }
+fn default_max_record_payload_bytes() -> usize { 256 * 1024 }
+
+impl Default for InfoConfiguration {
+    #[inline]
+    fn default() -> InfoConfiguration {
+        InfoConfiguration {
+            max_request_bytes: default_max_request_bytes(),
+            max_record_payload_bytes: default_max_record_payload_bytes(),
+            max_post_records: usize::max_value(),
+            max_post_bytes: usize::max_value(),
+            max_total_records: usize::max_value(),
+            max_total_bytes: usize::max_value(),
+        }
+    }
+}
+
+#[derive(Debug, Clone, Deserialize)]
+pub struct UploadResult {
+    batch: Option<String>,
+    /// Maps record id => why failed
+    #[serde(default = "HashMap::new")]
+    pub failed: HashMap<String, String>,
+    /// Vec of ids
+    #[serde(default = "Vec::new")]
+    pub success: Vec<String>
+}
+
+// Easier to fake during tests
+#[derive(Debug, Clone)]
+pub struct PostResponse {
+    pub status: StatusCode,
+    pub result: UploadResult, // This is lazy...
+    pub last_modified: ServerTimestamp,
+}
+
+impl PostResponse {
+    pub fn from_response(r: &mut Response) -> Result<PostResponse> {
+        let result: UploadResult = r.json()?;
+        // TODO Can this happen in error cases?
+        let last_modified = r.headers().get::<XLastModified>().map(|h| **h).ok_or_else(||
+            error::unexpected("Server didn't send X-Last-Modified header"))?;
+        let status = r.status();
+        Ok(PostResponse { status, result, last_modified })
+    }
+}
+
+
+#[derive(Debug, Clone, Eq, PartialEq, Ord, PartialOrd, Hash)]
+pub enum BatchState {
+    Unsupported,
+    NoBatch,
+    InBatch(String),
+}
+
+#[derive(Debug)]
+pub struct PostQueue<Post, OnResponse> {
+    poster: Post,
+    // XXX: It's cludgey that this needs to be pub just so we can get at the data it saves...
```

Comment:
for my education, why is this cludgy and what would be a better way to express it?